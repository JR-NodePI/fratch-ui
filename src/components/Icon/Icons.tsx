export { default as IconArrowDown } from './IconArrowDown';
export { default as IconClose } from './IconClose';
export { default as IconPlus } from './IconPlus';
export { default as IconCheck } from './IconCheck';
