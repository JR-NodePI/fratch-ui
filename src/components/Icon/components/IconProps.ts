export const IconType = Object.freeze({
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  ERROR: 'error',
  INFO: 'info',
  SUCCESS: 'success',
  WARNING: 'warning',
});

export type IconProps = {
  svg: React.ReactElement<JSX.IntrinsicElements['svg']>;
  className?: string;
  type?: (typeof IconType)[keyof typeof IconType];
};
