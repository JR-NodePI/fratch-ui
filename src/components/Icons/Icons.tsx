export { default as IconArrowDown } from './Icons/IconArrowDown';
export { default as IconCheck } from './Icons/IconCheck';
export { default as IconClose } from './Icons/IconClose';
export { default as IconDownload } from './Icons/IconDownload';
export { default as IconEdit } from './Icons/IconEdit';
export { default as IconError } from './Icons/IconError';
export { default as IconExport } from './Icons/IconExport';
export { default as IconImport } from './Icons/IconImport';
export { default as IconInfo } from './Icons/IconInfo';
export { default as IconPause } from './Icons/IconPause';
export { default as IconPin } from './Icons/IconPin';
export { default as IconPlay } from './Icons/IconPlay';
export { default as IconPlus } from './Icons/IconPlus';
export { default as IconSuccess } from './Icons/IconSuccess';
export { default as IconUpload } from './Icons/IconUpload';
export { default as IconVerticalDots } from './Icons/IconVerticalDots';
export { default as IconWarning } from './Icons/IconWarning';
