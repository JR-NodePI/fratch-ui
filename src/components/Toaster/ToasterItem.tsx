import { useEffect, useState } from 'react';

import { c } from '../../helpers/classNameHelpers';
import ButtonCloser from '../ButtonCloser/ButtonCloser';
import { IconError, IconInfo, IconSuccess, IconWarning } from '../Icons/Icons';
import { ToasterType } from './ToasterConstants';
import { nlToNodes } from './ToasterFormatHelpers';
import { type Toaster } from './ToasterListContextProps';

import styles from './Toaster.module.css';

const renderIconByType = (type: string): JSX.Element => {
  switch (type) {
    case ToasterType.SUCCESS:
      return <IconSuccess type={type} className={c(styles.icon)} />;
    case ToasterType.ERROR:
      return <IconError type={type} className={c(styles.icon)} />;
    case ToasterType.WARNING:
      return <IconWarning type={type} className={c(styles.icon)} />;
    case ToasterType.INFO:
      return <IconInfo type={type} className={c(styles.icon)} />;
    default:
      return <></>;
  }
};

export default function ToasterItem({
  onClose,
  toaster,
  className,
}: {
  onClose: (id: string) => void;
  toaster: Toaster;
  className?: string;
}): JSX.Element {
  const { duration, id, message, nlToBr, title, type } = toaster;

  const [mustClose, setMustClose] = useState<boolean>(false);
  const [cssClassStatus, setCssClassStatus] = useState<string>('');

  useEffect(() => {
    setCssClassStatus(styles.open);
  }, []);

  useEffect(() => {
    let timeoutId: ReturnType<typeof setTimeout>;
    if (mustClose && id) {
      setCssClassStatus(styles.close);
      timeoutId = setTimeout(() => {
        onClose(id);
      }, 300);
    }
    return () => {
      clearTimeout(timeoutId);
    };
  }, [mustClose, id, onClose]);

  useEffect(() => {
    let timeoutId: ReturnType<typeof setTimeout>;
    if (duration) {
      timeoutId = setTimeout(() => {
        setMustClose(true);
      }, duration);
    }
    return () => {
      clearTimeout(timeoutId);
    };
  }, [duration]);

  const handleClose = () => {
    setMustClose(true);
  };

  const handleRef = (node: HTMLDivElement | null): void => {
    if (!node) {
      return;
    }

    const itemRect = node?.getBoundingClientRect();

    if (itemRect) {
      node.style.setProperty('height', `${itemRect.height}px`);
    }
  };

  const finalMessage = nlToBr && message ? nlToNodes(message) : message;
  const toMuchDuration = !duration || duration >= 3000;

  return (
    <div
      key={id}
      ref={handleRef}
      className={c(
        styles.toaster_item,
        styles[type],
        className,
        cssClassStatus
      )}
    >
      {renderIconByType(type)}
      {title && <h5 className={c(styles.title)}>{title}</h5>}
      {message && <p className={c(styles.message)}>{finalMessage}</p>}
      {toMuchDuration && <ButtonCloser onClick={handleClose} />}
    </div>
  );
}
