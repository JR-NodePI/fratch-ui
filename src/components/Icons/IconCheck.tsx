import IconContainer from './components/IconContainer';
import { IconProps } from './components/IconProps';

const IconCheck = ({
  type,
  className,
}: Pick<IconProps, 'type' | 'className'>): JSX.Element => {
  return (
    <IconContainer className={className} type={type}>
      {({ iconClassName }) => (
        <svg
          className={iconClassName}
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="-1 -3 18 18"
        >
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2.2"
            d="M1 5.917 5.724 10.5 15 1.5"
          />
        </svg>
      )}
    </IconContainer>
  );
};

export default IconCheck;
