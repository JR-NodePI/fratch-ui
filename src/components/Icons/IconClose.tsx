import IconContainer from './components/IconContainer';
import { IconProps } from './components/IconProps';

const IconClose = ({
  type,
  className,
}: Pick<IconProps, 'type' | 'className'>): JSX.Element => {
  return (
    <IconContainer className={className} type={type}>
      {({ iconClassName }) => (
        <svg
          className={iconClassName}
          aria-hidden="true"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="-1 -1 16 16"
        >
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
          />
        </svg>
      )}
    </IconContainer>
  );
};

export default IconClose;
