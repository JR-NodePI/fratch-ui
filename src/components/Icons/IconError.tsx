import IconContainer from './components/IconContainer';
import { IconProps } from './components/IconProps';

const IconError = ({
  type,
  className,
}: Pick<IconProps, 'type' | 'className'>): JSX.Element => {
  return (
    <IconContainer className={className} type={type}>
      {({ iconClassName }) => (
        <svg
          className={iconClassName}
          aria-hidden="true"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 20 20"
        >
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="m13 7-6 6m0-6 6 6m6-3a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
          />
        </svg>
      )}
    </IconContainer>
  );
};

export default IconError;
