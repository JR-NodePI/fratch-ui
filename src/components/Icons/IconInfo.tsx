import IconContainer from './components/IconContainer';
import { IconProps } from './components/IconProps';

const IconInfo = ({
  type,
  className,
}: Pick<IconProps, 'type' | 'className'>): JSX.Element => {
  return (
    <IconContainer
      className={className}
      svg={
        <svg
          aria-hidden="true"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 20 20"
        >
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="M8 9h2v5m-2 0h4M9.408 5.5h.01M19 10a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z"
          />
        </svg>
      }
      type={type}
    />
  );
};

export default IconInfo;
