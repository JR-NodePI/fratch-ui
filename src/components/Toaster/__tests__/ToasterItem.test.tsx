import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import MockComponent from '../../../__mocks__/MockComponent';
import * as Icons from '../../Icons/Icons';
import { IconProps } from '../../Icons/Icons/IconProps';
import ToasterItem from '../ToasterItem';
import { Toaster, ToasterItemProps } from '../ToasterProps';

describe('ToasterItem', () => {
  const onClose = vi.fn();
  const initialToaster: Toaster = { type: 'info', duration: 100 };
  const initialProps = { onClose, toaster: initialToaster };

  const setup = (
    props: ToasterItemProps = initialProps
  ): ReturnType<typeof render> => {
    const { container, ...data } = render(<ToasterItem {...props} />);
    return {
      container,
      ...data,
    };
  };

  beforeEach(() => {
    vi.spyOn(Icons, 'IconClose').mockImplementation((props: IconProps) => (
      <MockComponent __testDisplayName="IconClose" {...props} />
    ));
    vi.spyOn(Icons, 'IconSuccess').mockImplementation((props: IconProps) => (
      <MockComponent __testDisplayName="IconSuccess" {...props} />
    ));
    vi.spyOn(Icons, 'IconError').mockImplementation((props: IconProps) => (
      <MockComponent __testDisplayName="IconError" {...props} />
    ));
    vi.spyOn(Icons, 'IconWarning').mockImplementation((props: IconProps) => (
      <MockComponent __testDisplayName="IconWarning" {...props} />
    ));
    vi.spyOn(Icons, 'IconInfo').mockImplementation((props: IconProps) => (
      <MockComponent __testDisplayName="IconInfo" {...props} />
    ));
    vi.spyOn(Icons, 'IconPin').mockImplementation((props: IconProps) => (
      <MockComponent __testDisplayName="IconPin" {...props} />
    ));
  });

  afterEach(() => {
    vi.restoreAllMocks();
    onClose.mockClear();
  });

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with type success', () => {
    const toaster: Toaster = { ...initialToaster, type: 'success' };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render with type error', () => {
    const toaster: Toaster = { ...initialToaster, type: 'error' };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render with type warning', () => {
    const toaster: Toaster = { ...initialToaster, type: 'warning' };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render with type nlToBr', () => {
    const message = `
      line 1
            line 2`;
    const toaster: Toaster = { ...initialToaster, message, nlToBr: true };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render with title', () => {
    const toaster: Toaster = { ...initialToaster, title: 'mock title' };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render with message', () => {
    const toaster: Toaster = { ...initialToaster, message: 'mock message' };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render when is stoppable with duration 0', () => {
    const toaster: Toaster = {
      ...initialToaster,
      duration: 0,
      stoppable: true,
    };
    const { container } = setup({ ...initialProps, toaster });
    expect(container).toMatchSnapshot();
  });

  it('should render when is stoppable with duration is greater than 2999 ms', async () => {
    const toaster: Toaster = {
      ...initialToaster,
      duration: 3000,
      stoppable: true,
    };
    const { container } = setup({ ...initialProps, toaster });

    await userEvent.click(screen.getByRole('button', { name: /pin/ }));

    expect(container).toMatchSnapshot();
  });

  it('should render when is stoppable but duration is 2999 ms', async () => {
    const toaster: Toaster = {
      ...initialToaster,
      duration: 2999,
      stoppable: true,
    };
    const { container } = setup({ ...initialProps, toaster });

    expect(container).toMatchSnapshot();
  });

  it('should render when is stoppable with duration 0 and was pinned', async () => {
    const toaster: Toaster = {
      ...initialToaster,
      duration: 0,
      stoppable: true,
    };
    const { container } = setup({ ...initialProps, toaster });

    await userEvent.click(screen.getByRole('button', { name: /pin/ }));

    expect(container).toMatchSnapshot();
  });

  it('should perform onClose with duration 0', async () => {
    const toasterId = 'mock-id';
    const toaster: Toaster = { type: 'info', duration: 0, id: toasterId };

    const { container } = setup({ ...initialProps, toaster });

    await userEvent.click(screen.getByRole('button'));

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith(toasterId);
    });

    expect(container).toMatchSnapshot();
  });

  it('should perform closer after duration', async () => {
    const toasterId = 'mock-id';
    const toaster: Toaster = { type: 'info', duration: 100, id: toasterId };

    const { container } = setup({ ...initialProps, toaster });

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith(toasterId);
    });

    expect(container).toMatchSnapshot();
  });
});
