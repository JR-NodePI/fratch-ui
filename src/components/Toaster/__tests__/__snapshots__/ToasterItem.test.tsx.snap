// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToasterItem > should perform closer after duration 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _close_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
  </div>
</div>
`;

exports[`ToasterItem > should perform onClose with duration 0 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _close_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
    <button
      class="_button_closer_8d9a3f"
      title="close"
    >
      <code
        aria-label="IconClose props"
      >
        className : _icon_8d9a3f
      </code>
      <i>
        close
      </i>
    </button>
  </div>
</div>
`;

exports[`ToasterItem > should render properly 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
  </div>
</div>
`;

exports[`ToasterItem > should render when is stoppable but duration is 2999 ms 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
  </div>
</div>
`;

exports[`ToasterItem > should render when is stoppable with duration 0 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
    <button
      class="_pin_button_68dafd"
    >
      <code
        aria-label="IconPin props"
      >
        className : _pin_button_icon_68dafd
      </code>
      <i>
        pin
      </i>
    </button>
    <button
      class="_button_closer_8d9a3f"
      title="close"
    >
      <code
        aria-label="IconClose props"
      >
        className : _icon_8d9a3f
      </code>
      <i>
        close
      </i>
    </button>
  </div>
</div>
`;

exports[`ToasterItem > should render when is stoppable with duration 0 and was pinned 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
    <button
      class="_pin_button_68dafd _pinned_68dafd"
    >
      <code
        aria-label="IconPin props"
      >
        className : _pin_button_icon_68dafd
      </code>
      <i>
        pin
      </i>
    </button>
    <button
      class="_button_closer_8d9a3f"
      title="close"
    >
      <code
        aria-label="IconClose props"
      >
        className : _icon_8d9a3f
      </code>
      <i>
        close
      </i>
    </button>
  </div>
</div>
`;

exports[`ToasterItem > should render when is stoppable with duration is greater than 2999 ms 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
    <button
      class="_pin_button_68dafd _pinned_68dafd"
    >
      <code
        aria-label="IconPin props"
      >
        className : _pin_button_icon_68dafd
      </code>
      <i>
        pin
      </i>
    </button>
    <button
      class="_button_closer_8d9a3f"
      title="close"
    >
      <code
        aria-label="IconClose props"
      >
        className : _icon_8d9a3f
      </code>
      <i>
        close
      </i>
    </button>
  </div>
</div>
`;

exports[`ToasterItem > should render with message 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
    <p
      class="_message_68dafd"
    >
      mock message
    </p>
  </div>
</div>
`;

exports[`ToasterItem > should render with title 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _info_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconInfo props"
    >
      type : info
      className : _icon_68dafd
    </code>
    <h5
      class="_title_68dafd"
    >
      mock title
    </h5>
  </div>
</div>
`;

exports[`ToasterItem > should render with type error 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _error_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconError props"
    >
      type : error
      className : _icon_68dafd
    </code>
  </div>
</div>
`;

exports[`ToasterItem > should render with type success 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _success_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconSuccess props"
    >
      type : success
      className : _icon_68dafd
    </code>
  </div>
</div>
`;

exports[`ToasterItem > should render with type warning 1`] = `
<div>
  <div
    class="_toaster_item_68dafd _warning_68dafd _open_68dafd"
    style="height: 0px;"
  >
    <code
      aria-label="IconWarning props"
    >
      type : warning
      className : _icon_68dafd
    </code>
  </div>
</div>
`;
